syntax = "proto2";

package ad_proto;

option cc_generic_services = true;

//------------------------------------------------

// 广告信息
message AdInfo {
    optional uint64 id = 1;
    optional string imageUrl = 2;      // 图片链接
    optional string url = 3;            // 连接
    optional string content = 4;
    optional uint32 type = 5;       // 投放类型
}
message GetAdInfoReq {
    optional uint32 type = 1;   // 广告类型
    optional uint32 num = 2;    // 数量
}
message GetAdInfoResp {
    optional uint32 err = 1;
    optional string msg = 2;
    repeated AdInfo info = 3; 
}

// CPC单次点击
message CostPerClickReq {
    optional uint64 id = 1;
}
message CostPerClickResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// CPM千次曝光广告
message CostPerMilleReq {
    optional uint64 id = 1;
}
message CostPerMilleResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// CPA行动收费,浏览15s
message CostPerVisitReq {
    optional uint64 id = 1;
}
message CostPerVisitResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// CPA，加入购物车收费
message CostPerShopReq {
    optional uint64 id = 1;
}
message CostPerShopResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// CPT,按时间收费
message CostPerTimeReq {
    optional uint64 id = 1;
}

message CostPerTimeResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// CPS,按销售收费
message CostPerSellReq {
    optional uint64 id = 1;
}

message CostPerSellResp {
    optional uint32 err = 1;
    optional string msg = 2;
}

// 广告投放
message PutAdvertiseReq {
    optional string uuid = 1;
    optional string imageUrl = 2;   // 图片url
    optional string url = 3;        // url
    optional string content = 4;    // 内容描述
    optional uint32 type = 5;       // 投放类型
}
message PutAdvertiseResp {
    optional uint32 err = 1;
    optional string msg = 2;
    optional AdInfo info = 3;
}

//-----------------------------------------------

// 注册
message RegisterReq {
    optional string username = 1;
    optional string password = 2;
    optional uint64 phone = 3;
}

message RegisterResp {
    optional uint32 err = 1;
    optional string msg = 2;
    optional string uuid = 3;   // 用户唯一标识
}

// 登录
message LoginReq {
    optional string username = 1;
    optional string password = 2;
}

message LoginResp {
    optional uint32 err = 1;
    optional string msg = 2;
    optional string uuid = 3;   // 用户唯一标识
    optional uint64 phone = 4;
}
// ----------------------------------------------------------------

// 获得所有信息
message GetFundInfoReq {
    optional string uuid = 1;
}
message GetFundInfoResp {
    optional int32 err = 1;
    optional string msg = 2;
    optional string uuid = 3;
    optional double amount = 4;//总金额
    optional double welfare = 5;//红包福利
}

// 充值
message RechargeReq {
    optional string uuid = 1;
    optional double money = 2;
}
message RechargeResp {
    optional int32 err = 1;
    optional string msg = 2;
    optional string uuid = 3;
    optional double amount = 4;//总金额
    optional double welfare = 5;//红包福利
}

// 扣费
message DeductionReq {
    optional string uuid = 1;
    optional double money = 2;
}
message DeductionResp {
    optional int32 err = 1;
    optional string msg = 2;
    optional string uuid = 3;
    optional double amount = 4;//总金额
    optional double welfare = 5;//红包福利
}

// 获取统计
message GetCountReq {
    optional string uuid = 1;
}

message GetCountResp {
    optional uint32 err = 1;
    optional string msg = 2;
    optional double costs = 3;
    optional uint64 clickNum = 4;   // 点击
    optional uint64 showNum = 5;    // 展示
    optional uint64 sellNum = 6;    // 销售
    optional uint64 visitNum = 7; // 浏览
    optional uint64 shopNum = 8;  // 加购
    optional double clickCost = 9;
    optional double showCost = 10;
}

//-----------------------------------------------
message HttpRequest {

};
message HttpResponse {

};

service HttpService {
    rpc CostPerClick (HttpRequest) returns (HttpResponse);   // CPC单次点击
    rpc CostPerMille (HttpRequest) returns (HttpResponse);   // CPM千次曝光广告
    rpc CostPerVisit (HttpRequest) returns (HttpResponse);   // CPA行动收费，浏览15s
    rpc CostPerShop (HttpRequest) returns (HttpResponse);   // CPA行动收费,加购
    rpc CostPerTime (HttpRequest) returns (HttpResponse);   // CPT时间收费
    rpc CostPerSell (HttpRequest) returns (HttpResponse);   // CPS销售收费
    rpc PutAdvertise (HttpRequest) returns (HttpResponse);   // 广告投放
    rpc GetAdInfo (HttpRequest) returns (HttpResponse);   // 广告投放

    rpc GetFundInfo (HttpRequest) returns (HttpResponse);  // 获得所有信息
    rpc Recharge (HttpRequest) returns (HttpResponse);     // 充值
    rpc Deduction (HttpRequest) returns (HttpResponse);     // 扣费
    rpc GetCount (HttpRequest) returns (HttpResponse);     // 获取统计

    rpc Register (HttpRequest) returns (HttpResponse);  // 注册
    rpc Login (HttpRequest) returns (HttpResponse);     // 登录
}